api/src/main/java/org/openmrs/module/registrationcore/api/impl/PatientCreatedListener.java

```diff
@@ -0,0 +1,129 @@
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This class listens for patient CREATED events.
+ * If MPI is enabled, it exports patient to MPI,
+ * and updates the local patient with the new patient identifier generated by the MPI.
+ */
+public class PatientCreatedListener extends PatientActionListener {
+    
+    private static final Logger LOGGER = LoggerFactory.getLogger(PatientCreatedListener.class);
+    
+    private MpiProperties mpiProperties;
+
+    private IdentifierBuilder identifierBuilder;
+
+    public void setMpiProperties(MpiProperties mpiProperties) {
+        this.mpiProperties = mpiProperties;
+    }
+
+    public void setIdentifierBuilder(IdentifierBuilder identifierBuilder) {
+        this.identifierBuilder = identifierBuilder;
+    }
+
+    /**
+     * Defines the list of Actions that this subscribable event listener class listens out for.
+     *
+     * @return a list of Actions this listener can deal with
+     */
+    public List<String> subscribeToActions(){
+        List actions = new ArrayList<String>();
+        actions.add(Event.Action.CREATED.name());
+        return actions;
+    }
+
+    /**
+     * Exports patient to MPI, and updates the local patient
+     * with the new patient identifier generated by the MPI.
+     *
+     * @param message message with properties.
+     */
+    @Override
+    public void performMpiAction(Message message) {
+        Patient patient = extractPatient(message);
+        if (patient != null){
+            PatientIdentifierType mpiPatientIdType = patientService.getPatientIdentifierTypeByUuid(mpiProperties.getMpiPersonIdentifierTypeUuid());
+            PatientIdentifier mpiPatientId = patient.getPatientIdentifier(mpiPatientIdType);
+            if (mpiPatientId != null && !mpiPatientId.getIdentifier().isEmpty()) {
+                // Patient originated in MPI and thus should not be pushed to the MPI
+                // TODO does mpi have openmrs id, if no update, if yes stop
+                return;
+            }else{
+                String mpiPatientIdString = null;
+                try {
+                    mpiPatientIdString = pushPatientToMpiServer(patient);
+                } catch (Exception e) {
+                    ErrorHandlingService errorHandler = coreProperties.getPixErrorHandlingService();
+                    if (errorHandler == null) {
+                        throw new MpiException("PIX patient push exception occurred "
+                                + "with not configured PIX error handler", e);
+                    } else {
+                        errorHandler.handle(prepareParameters(patient),
+                                PixErrorHandlingService.SENDING_PATIENT_AFTER_PATIENT_CREATION_DESTINATION,
+                                true,
+                                ExceptionUtils.getFullStackTrace(e));
+                        throw new MpiException("PIX patient push exception occurred", e);
```

Comment:
If you throw an exception, won't it result in a rollback if this is happening inside a transaction?

Reply:
Sorry could you explain what you mean and the implications in more detail thanks ><

Reply:
I guess it isn't an issue since the listener isn't called from a transactional method.